File: Analyzer.Backend/Core/Result.cs
using System.Net;

namespace Analyzer.Backend.Core;

public readonly struct Result<T>
{
    private readonly T? value;
    private readonly Error? error;

    private Result(T value)
    {
        this.value = value;
        error = null;
        IsSuccess = true;
    }

    private Result(Error error)
    {
        value = default;
        this.error = error;
        IsSuccess = false;
    }

    public bool IsSuccess { get; }

    public T Value
        => IsSuccess ? value! : throw new InvalidOperationException($"Cannot access Value when Result is failure. Error: {error}");

    public Error Error => !IsSuccess ? error! : throw new InvalidOperationException("Cannot access Error when Result is success.");

    public static Result<T> Success(T value) => new(value);

    public static Result<T> Failure(Error error) => new(error);

    public static Result<T> Failure(string message, HttpStatusCode code) => new(new Error(message, code));

    public TResult Match<TResult>(Func<T, TResult> onSuccess, Func<Error, TResult> onFailure)
        => IsSuccess ? onSuccess(value!) : onFailure(error!);

    public Result<T> OnFailure(Action<Error> action)
    {
        if (!IsSuccess)
        {
            action(error!);
        }

        return this;
    }

    public Result<T> OnSuccess(Action<T> action)
    {
        if (IsSuccess)
        {
            action(value!);
        }

        return this;
    }
}

public record Error(string Message, HttpStatusCode? Code = null)
{
    public override string ToString() => Code is not null ? $"[{Code}] {Message}" : $"{Message}";
}

File: src/Warehouse.Core.Functional/Shared/Result.fs
namespace Warehouse.Core.Functional.Shared

open System
open System.Net

type Error(message: string, code: Nullable<HttpStatusCode>) =
    new(message: string) = Error(message, Nullable())
    member this.Message = message
    member this.Code = code
    override this.ToString() = if code.HasValue then $"[{code.Value}] %s{message}" else message

[<Struct>]
type Result<'T> =
    val private value: 'T
    val private error: Error
    val public IsSuccess: bool

    private new(value: 'T) = { value = value; error = null; IsSuccess = true }
    private new(error: Error) = { value = Unchecked.defaultof<'T>; error = error; IsSuccess = false }

    member this.Value =
        if this.IsSuccess then
            this.value
        else
            raise (InvalidOperationException $"Cannot access Value when Result is failure. Error: {this.error}")

    member this.Error =
        if not this.IsSuccess then
            this.error
        else
            raise (InvalidOperationException("Cannot access Error when Result is success."))

    static member Success(value: 'T) = Result<'T>(value)
    static member Failure(error: Error) = Result<'T>(error)
    static member Failure(message: string, code: HttpStatusCode) = Result<'T>(Error(message, Nullable(code)))
    static member Failure(message: string) = Result<'T>(Error(message))

    member this.Match<'TResult>(onSuccess: Func<'T, 'TResult>, onFailure: Func<Error, 'TResult>) =
        if this.IsSuccess then onSuccess.Invoke(this.value) else onFailure.Invoke(this.error)

    member this.OnFailure(action: Action<Error>) =
        if not this.IsSuccess then
            action.Invoke(this.error)

        this

    member this.OnSuccess(action: Action<'T>) =
        if this.IsSuccess then
            action.Invoke(this.value)

        this
